RPIPs/RPIP-onchain-pdao.md

```diff
@@ -0,0 +1,457 @@
+is the challenged index and a merkle proof from the challenged index back to the proposal root. A proposer will be
+unable to respond if the response requires submitting a pollard containing leaf nodes where the leaf nodes do not match
+the actual values on chain. In this way, they are forced to not respond for `proposal.challenge.period` time. After
+which, the proposal can be defeated.
+
+If a proposal is defeated, the proposer forfeits their bond which is divided proportionally amongst the challengers who
+contributed to the proposal's defeat.
+
+To raise a proposal, a node operator MUST
+call `propose(string proposalMessage, bytes _payload, uint32 blockNumber, Type.Node[] treeNodes)`
+on `RocketDAOProtocolProposals`. Where `proposalMessage` is an arbitrary string, `_payload` is an ABI encoded function
+call to `RocketDAOProtocolProposals`, `blockNumber` is the block number used to generate the merkle pollard,
+and `treeNodes` is the merkle pollard.
+
+To respond to a challenge, the proposer MUST
+call `submitRoot(uint256 proposalID, uint256 index, Types.Node[] witness, Types.Node[] nodes`
+on `RocketDAOProtocolVerifier`. Where `proposalID` is the proposal's ID, `index` is the challenged index, `witness` is
+the merkle proof and `nodes` is the new merkle pollard from the challenged index.
+
+To claim a refund for a proposal bond and rewards from any invalid challenges, the proposer MUST
+call `claimBondProposer(uint256 proposalID, uint256[] indices)` on `RocketDAOProtocolVerifier`. Where `proposalID` is
+the proposal's ID and `indices` is an array of indices that are being claimed. The index "1" refers to the proposal
+root. Claiming this index will return the proposal bond.
+
+### Verifying
+
+Any node MAY challenge a proposal by supplying an index into the merkle-sum tree that they are alleging is incorrect.
+The challenger MUST have at least `proposal.challenge.bond` unlocked effective RPL stake. And this amount of the
+challenger's RPL SHALL BE locked until the proposal is defeated.
+
+The index which is alleged to be incorrect MUST be an index of a node in one of the pollards that the proposer has
+submitted. Either the initial pollard submitted with the proposal, or any subsequent response pollards to a challenge.
+
+If the initial pollard submitted by the proposer is invalid, they will be unable to respond to challenges which require
+the proposer to provide information that is known on chain. It will take a few rounds of challenges and responses to
+reach the last layer of the merkle-sum tree. After which, the proposer will no longer be able to respond to challenges.
+
+After `proposal.challenge.period` has passed and the proposer has not responded to the challenge, any one MAY
+defeat the proposal. If a proposal is defeated, it SHALL NOT enter the voting period. And SHALL NOT be executed.
+
+After defeating a proposal, each challenger who participated in defeating the challenge SHALL be paid a proportional
```

Comment:
What does it mean to participate? To be one of the challenges on an unanswered branch? (this is answered later; either order needs to swap or we at least need a breadcrumb of "defined below" or something)

What does "proportional" mean? Is it weighted by challenge? Like there were 6 challenges involved, you did 3 of them, so you get half the payout?

Reply:
Hmmm... I think there's a problem vector here :/

The pollard submitter can get most of their money back. They get challenged. They realize they've been caught and so they submit their next pollard bundled with a challenge to themselves. Essentially they can ensure that they payouts for N-1 of N challenges. The only defense that comes to mind is providing a period where only the initial challenger can provided the next challenge (and opening it up generally afterwards).

Reply:
w.r.t proportional. Yes.

I don't think this is an issue. Ultimately the goal is to kill an invalid proposal. So if the proposer themselves wants to kill it then so be it.

Reply:
🤔 I see two goals:
- kill the fraudulent proposal (safety this time) 
- punish the fraudulent proposer (discourage attacks) 

I think I still like a small period reserved for the initial challenger to continue.